src/App.tsx

```diff
@@ -15,35 +16,101 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
+const SORT_FIELD_ALPHABET = 'alphabet';
+const SORT_FIELD_LENGTH = 'length';
+
+interface FilterParams<T> {
+  sort: keyof T | '';
+  reversed: boolean;
+}
+
+function getPrepearedGoods<T>(
+  goods: string[],
+  { sort, reversed }: FilterParams<T>,
+): string[] {
+  const preparedGoods = [...goods];
+
+  if (sort) {
+    preparedGoods.sort((goods1, goods2) => {
+      switch (sort) {
+        case SORT_FIELD_ALPHABET:
+          return goods1.localeCompare(goods2);
+        case SORT_FIELD_LENGTH:
+          return goods1.length - goods2.length;
+        default:
+          return 0;
+      }
+    });
+
+    if (reversed) {
+      return preparedGoods.reverse();
+    }
+  }
+
+  return preparedGoods;
+}
+
 export const App: React.FC = () => {
+  const [sort, setSort] = useState('');
+  const [reversed, setReversed] = useState(false);
+
+  const visibleGoods = getPrepearedGoods(goodsFromServer, {
+    sort,
+    reversed,
+  });
+
   return (
     <div className="section content">
       <div className="buttons">
-        <button type="button" className="button is-info is-light">
+        <button
+          type="button"
+          className={cn('button', 'is-info', {
+            'is-light': sort !== SORT_FIELD_ALPHABET,
+          })}
+          onClick={() => setSort(SORT_FIELD_ALPHABET)}
+        >
           Sort alphabetically
         </button>
 
-        <button type="button" className="button is-success is-light">
+        <button
+          type="button"
+          className={cn('button', 'is-info', {
+            'is-light': sort !== SORT_FIELD_LENGTH,
+          })}
+          onClick={() => setSort(SORT_FIELD_LENGTH)}
+        >
           Sort by length
         </button>
 
-        <button type="button" className="button is-warning is-light">
+        <button
+          type="button"
+          className={cn('button', 'is-info', {
+            'is-light': reversed !== false,
+          })}
+          onClick={() => setReversed(!reversed)}
+        >
           Reverse
         </button>
 
-        <button type="button" className="button is-danger is-light">
-          Reset
-        </button>
+        {(sort !== '' || reversed !== false) && (
+          <button
+            type="button"
+            className="button is-danger is-light"
+            onClick={() => {
+              setSort('');
+              setReversed(false);
+            }}
+          >
+            Reset
+          </button>
+        )}
       </div>
 
       <ul>
         <ul>
```

Comment:
The `<ul>` tag on this line is redundant and should be removed to avoid unnecessary nesting. The `visibleGoods.map` function already creates a list of `<li>` elements within a single `<ul>`.